File: NestCash/frontend/lib/models/challenge.dart
// lib/models/challenge.dart
class ChallengeReward {
  final int points;
  final List<String> badges;
  final String? title;

  ChallengeReward({
    required this.points,
    required this.badges,
    this.title,
  });

  factory ChallengeReward.fromJson(Map<String, dynamic> json) {
    return ChallengeReward(
      points: json['points'] ?? 0,
      badges: List<String>.from(json['badges'] ?? []),
      title: json['title'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'points': points,
      'badges': badges,
      'title': title,
    };
  }
}

class ChallengeRule {
  final String type;
  final double value;
  final String description;

  ChallengeRule({
    required this.type,
    required this.value,
    required this.description,
  });

  factory ChallengeRule.fromJson(Map<String, dynamic> json) {
    return ChallengeRule(
      type: json['type'],
      value: json['value'].toDouble(),
      description: json['description'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'type': type,
      'value': value,
      'description': description,
    };
  }
}

class ChallengeProgress {
  final double currentValue;
  final double targetValue;
  final String unit;
  final double percentage;

  ChallengeProgress({
    required this.currentValue,
    required this.targetValue,
    required this.unit,
    required this.percentage,
  });

  factory ChallengeProgress.fromJson(Map<String, dynamic> json) {
    return ChallengeProgress(
      currentValue: json['current_value']?.toDouble() ?? 0.0,
      targetValue: json['target_value']?.toDouble() ?? 0.0,
      unit: json['unit'] ?? 'HUF',
      percentage: json['percentage']?.toDouble() ?? 0.0,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'current_value': currentValue,
      'target_value': targetValue,
      'unit': unit,
      'percentage': percentage,
    };
  }
}

enum ChallengeType {
  savings('savings', 'Megtakarítás'),
  expenseReduction('expense_reduction', 'Kiadás csökkentés'),
  habitStreak('habit_streak', 'Szokás streak'),
  budgetControl('budget_control', 'Költségvetés betartás'),
  investment('investment', 'Befektetés'),
  incomeBoost('income_boost', 'Bevétel növelés');

  const ChallengeType(this.value, this.displayName);
  final String value;
  final String displayName;

  static ChallengeType fromString(String value) {
    return ChallengeType.values.firstWhere(
      (type) => type.value == value,
      orElse: () => ChallengeType.savings,
    );
  }
}

enum ChallengeDifficulty {
  easy('easy', 'Könnyű'),
  medium('medium', 'Közepes'),
  hard('hard', 'Nehéz'),
  expert('expert', 'Szakértő');

  const ChallengeDifficulty(this.value, this.displayName);
  final String value;
  final String displayName;

  static ChallengeDifficulty fromString(String value) {
    return ChallengeDifficulty.values.firstWhere(
      (difficulty) => difficulty.value == value,
      orElse: () => ChallengeDifficulty.easy,
    );
  }
}

enum ChallengeStatus {
  draft('draft', 'Tervezet'),
  active('active', 'Aktív'),
  completed('completed', 'Befejezett'),
  cancelled('cancelled', 'Megszakított');

  const ChallengeStatus(this.value, this.displayName);
  final String value;
  final String displayName;

  static ChallengeStatus fromString(String value) {
    return ChallengeStatus.values.firstWhere(
      (status) => status.value == value,
      orElse: () => ChallengeStatus.active,
    );
  }
}

enum ParticipationStatus {
  active('active', 'Aktív'),
  completed('completed', 'Befejezett'),
  failed('failed', 'Sikertelen'),
  abandoned('abandoned', 'Elhagyott');

  const ParticipationStatus(this.value, this.displayName);
  final String value;
  final String displayName;

  static ParticipationStatus fromString(String value) {
    return ParticipationStatus.values.firstWhere(
      (status) => status.value == value,
      orElse: () => ParticipationStatus.active,
    );
  }
}

class Challenge {
  final String id;
  final String title;
  final String description;
  final String? shortDescription;
  final ChallengeType challengeType;
  final ChallengeDifficulty difficulty;
  final int durationDays;
  final double? targetAmount;
  final List<ChallengeRule> rules;
  final ChallengeReward rewards;
  final ChallengeStatus status;
  final DateTime createdAt;
  final DateTime updatedAt;
  final int participantCount;
  final double completionRate;
  final String? imageUrl;
  final List<String> tags;
  final String? creatorUsername;
  final bool isParticipating;
  final ChallengeProgress? myProgress;
  final ParticipationStatus? myStatus;

  Challenge({
    required this.id,
    required this.title,
    required this.description,
    this.shortDescription,
    required this.challengeType,
    required this.difficulty,
    required this.durationDays,
    this.targetAmount,
    required this.rules,
    required this.rewards,
    required this.status,
    required this.createdAt,
    required this.updatedAt,
    required this.participantCount,
    required this.completionRate,
    this.imageUrl,
    required this.tags,
    this.creatorUsername,
    required this.isParticipating,
    this.myProgress,
    this.myStatus,
  });

  factory Challenge.fromJson(Map<String, dynamic> json) {
    return Challenge(
      id: json['id'],
      title: json['title'],
      description: json['description'],
      shortDescription: json['short_description'],
      challengeType: ChallengeType.fromString(json['challenge_type']),
      difficulty: ChallengeDifficulty.fromString(json['difficulty']),
      durationDays: json['duration_days'],
      targetAmount: json['target_amount']?.toDouble(),
      rules: (json['rules'] as List<dynamic>)
          .map((rule) => ChallengeRule.fromJson(rule))
          .toList(),
      rewards: ChallengeReward.fromJson(json['rewards']),
      status: ChallengeStatus.fromString(json['status']),
      createdAt: DateTime.parse(json['created_at']),
      updatedAt: DateTime.parse(json['updated_at']),
      participantCount: json['participant_count'],
      completionRate: json['completion_rate']?.toDouble() ?? 0.0,
      imageUrl: json['image_url'],
      tags: List<String>.from(json['tags'] ?? []),
      creatorUsername: json['creator_username'],
      isParticipating: json['is_participating'] ?? false,
      myProgress: json['my_progress'] != null 
          ? ChallengeProgress.fromJson(json['my_progress'])
          : null,
      myStatus: json['my_status'] != null 
          ? ParticipationStatus.fromString(json['my_status'])
          : null,
    );
  }
}

class UserChallenge {
  final String id;
  final String userId;
  final String username;
  final String challengeId;
  final ParticipationStatus status;
  final DateTime joinedAt;
  final DateTime? startedAt;
  final DateTime? completedAt;
  final DateTime updatedAt;
  final ChallengeProgress progress;
  final double? personalTarget;
  final String? notes;
  final int earnedPoints;
  final List<String> earnedBadges;
  final int bestStreak;
  final int currentStreak;
  final String challengeTitle;
  final ChallengeType challengeType;
  final ChallengeDifficulty challengeDifficulty;

  UserChallenge({
    required this.id,
    required this.userId,
    required this.username,
    required this.challengeId,
    required this.status,
    required this.joinedAt,
    this.startedAt,
    this.completedAt,
    required this.updatedAt,
    required this.progress,
    this.personalTarget,
    this.notes,
    required this.earnedPoints,
    required this.earnedBadges,
    required this.bestStreak,
    required this.currentStreak,
    required this.challengeTitle,
    required this.challengeType,
    required this.challengeDifficulty,
  });

  factory UserChallenge.fromJson(Map<String, dynamic> json) {
    return UserChallenge(
      id: json['id'],
      userId: json['user_id'],
      username: json['username'],
      challengeId: json['challenge_id'],
      status: ParticipationStatus.fromString(json['status']),
      joinedAt: DateTime.parse(json['joined_at']),
      startedAt: json['started_at'] != null 
          ? DateTime.parse(json['started_at'])
          : null,
      completedAt: json['completed_at'] != null 
          ? DateTime.parse(json['completed_at'])
          : null,
      updatedAt: DateTime.parse(json['updated_at']),
      progress: ChallengeProgress.fromJson(json['progress']),
      personalTarget: json['personal_target']?.toDouble(),
      notes: json['notes'],
      earnedPoints: json['earned_points'],
      earnedBadges: List<String>.from(json['earned_badges'] ?? []),
      bestStreak: json['best_streak'],
      currentStreak: json['current_streak'],
      challengeTitle: json['challenge_title'],
      challengeType: ChallengeType.fromString(json['challenge_type']),
      challengeDifficulty: ChallengeDifficulty.fromString(json['challenge_difficulty']),
    );
  }
}

File: NestCash/backend/app/models/challenge.py
# app/models/challenge.py
from __future__ import annotations
from beanie import Document, PydanticObjectId
from pydantic import BaseModel, Field
from typing import Optional, List, Dict, Any
from datetime import datetime
from enum import Enum

class ChallengeType(str, Enum):
    SAVINGS = "savings"  # Megtakarítás
    EXPENSE_REDUCTION = "expense_reduction"  # Kiadás csökkentés
    HABIT_STREAK = "habit_streak"  # Szokás streak
    BUDGET_CONTROL = "budget_control"  # Költségvetés betartás
    INVESTMENT = "investment"  # Befektetés
    INCOME_BOOST = "income_boost"  # Bevétel növelés

class ChallengeDifficulty(str, Enum):
    EASY = "easy"
    MEDIUM = "medium"
    HARD = "hard"
    EXPERT = "expert"

class ChallengeStatus(str, Enum):
    DRAFT = "draft"  # Tervezet
    ACTIVE = "active"  # Aktív
    COMPLETED = "completed"  # Befejezett
    CANCELLED = "cancelled"  # Megszakított

class ParticipationStatus(str, Enum):
    ACTIVE = "active"
    COMPLETED = "completed"
    FAILED = "failed"
    ABANDONED = "abandoned"

class ChallengeReward(BaseModel):
    points: int = Field(default=0, description="Pontok száma")
    badges: List[str] = Field(default_factory=list, description="Kitűzők listája")
    title: Optional[str] = Field(None, description="Különleges cím")

class ChallengeRule(BaseModel):
    type: str = Field(..., description="Szabály típusa (pl. 'min_amount', 'max_transactions')")
    value: float = Field(..., description="Szabály értéke")
    description: str = Field(..., description="Szabály leírása")

class ChallengeProgress(BaseModel):
    current_value: float = Field(default=0.0, description="Jelenlegi érték")
    target_value: float = Field(..., description="Cél érték")
    unit: str = Field(default="HUF", description="Mértékegység")
    percentage: float = Field(default=0.0, description="Százalékos teljesítmény")

# === CHALLENGE DOCUMENT ===
class ChallengeDocument(Document):
    title: str = Field(..., description="Kihívás címe")
    description: str = Field(..., description="Részletes leírás")
    short_description: Optional[str] = Field(None, description="Rövid leírás")
    
    # Alapvető tulajdonságok
    challenge_type: ChallengeType = Field(..., description="Kihívás típusa")
    difficulty: ChallengeDifficulty = Field(..., description="Nehézségi szint")
    duration_days: int = Field(..., description="Időtartam napokban", ge=1)
    
    # Célok és szabályok
    target_amount: Optional[float] = Field(None, description="Cél összeg (HUF)")
    rules: List[ChallengeRule] = Field(default_factory=list, description="Kihívás szabályai")
    
    # Jutalmak
    rewards: ChallengeReward = Field(default_factory=ChallengeReward, description="Jutalmak")
    
    # Metaadatok
    creator_id: Optional[PydanticObjectId] = Field(None, description="Létrehozó user ID (None = rendszer)")
    creator_username: Optional[str] = Field(None, description="Létrehozó felhasználónév")
    
    # Státusz és időbélyegek
    status: ChallengeStatus = Field(default=ChallengeStatus.ACTIVE, description="Kihívás státusza")
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)
    
    # Statisztikák
    participant_count: int = Field(default=0, description="Résztvevők száma")
    completion_rate: float = Field(default=0.0, description="Befejezési arány (%)")
    
    # Kép és kategorízálás
    image_url: Optional[str] = Field(None, description="Kihívás képe")
    tags: List[str] = Field(default_factory=list, description="Címkék")
    
    # Haladási tracking beállítások
    track_categories: List[str] = Field(default_factory=list, description="Nyomon követendő kategóriák")
    track_accounts: List[str] = Field(default_factory=list, description="Nyomon követendő számlák")
    
    class Settings:
        name = "challenges"
        indexes = [
            "challenge_type",
            "difficulty", 
            "status",
            "created_at",
            "participant_count",
            [("challenge_type", 1), ("difficulty", 1)],
            [("status", 1), ("created_at", -1)]
        ]

# === USER CHALLENGE PARTICIPATION ===
class UserChallengeDocument(Document):
    user_id: PydanticObjectId = Field(..., description="Felhasználó ID")
    username: str = Field(..., description="Felhasználónév")
    challenge_id: PydanticObjectId = Field(..., description="Kihívás ID")
    
    # Státuszok
    status: ParticipationStatus = Field(default=ParticipationStatus.ACTIVE)
    
    # Időbélyegek
    joined_at: datetime = Field(default_factory=datetime.utcnow)
    started_at: Optional[datetime] = Field(None, description="Tényleges kezdés ideje")
    completed_at: Optional[datetime] = Field(None)
    updated_at: datetime = Field(default_factory=datetime.utcnow)
    
    # Haladás követés
    progress: ChallengeProgress = Field(..., description="Haladás adatok")
    daily_progress: Dict[str, float] = Field(default_factory=dict, description="Napi haladás {YYYY-MM-DD: érték}")
    
    # Személyes célok és beállítások
    personal_target: Optional[float] = Field(None, description="Személyre szabott cél")
    notes: Optional[str] = Field(None, description="Személyes jegyzet")
    
    # Jutalmazás
    earned_points: int = Field(default=0, description="Szerzett pontok")
    earned_badges: List[str] = Field(default_factory=list, description="Szerzett kitűzők")
    
    # Statisztikák
    best_streak: int = Field(default=0, description="Legjobb sorozat (napokban)")
    current_streak: int = Field(default=0, description="Jelenlegi sorozat")
    
    class Settings:
        name = "user_challenges"
        indexes = [
            "user_id",
            "challenge_id",
            "status",
            "joined_at",
            [("user_id", 1), ("status", 1)],
            [("challenge_id", 1), ("status", 1)],
            [("user_id", 1), ("challenge_id", 1)]  # Unique constraint
        ]

# === PYDANTIC SCHEMAS ===

# Challenge schemas
class ChallengeCreate(BaseModel):
    title: str
    description: str
    short_description: Optional[str] = None
    challenge_type: ChallengeType
    difficulty: ChallengeDifficulty
    duration_days: int = Field(..., ge=1, le=365)
    target_amount: Optional[float] = None
    rules: List[ChallengeRule] = Field(default_factory=list)
    rewards: ChallengeReward = Field(default_factory=ChallengeReward)
    image_url: Optional[str] = None
    tags: List[str] = Field(default_factory=list)
    track_categories: List[str] = Field(default_factory=list)
    track_accounts: List[str] = Field(default_factory=list)

class ChallengeUpdate(BaseModel):
    title: Optional[str] = None
    description: Optional[str] = None
    short_description: Optional[str] = None
    status: Optional[ChallengeStatus] = None
    image_url: Optional[str] = None
    tags: Optional[List[str]] = None

class ChallengeRead(BaseModel):
    id: str
    title: str
    description: str
    short_description: Optional[str] = None
    challenge_type: ChallengeType
    difficulty: ChallengeDifficulty
    duration_days: int
    target_amount: Optional[float] = None
    rules: List[ChallengeRule]
    rewards: ChallengeReward
    status: ChallengeStatus
    created_at: datetime
    updated_at: datetime
    participant_count: int
    completion_rate: float
    image_url: Optional[str] = None
    tags: List[str]
    creator_username: Optional[str] = None
    
    # Felhasználó-specifikus mezők (csak ha be van jelentkezve a kihívásba)
    is_participating: bool = False
    my_progress: Optional[ChallengeProgress] = None
    my_status: Optional[ParticipationStatus] = None

class ChallengeListResponse(BaseModel):
    challenges: List[ChallengeRead]
    total_count: int
    skip: int
    limit: int

# User Challenge schemas
class UserChallengeJoin(BaseModel):
    personal_target: Optional[float] = None
    notes: Optional[str] = None

class UserChallengeUpdate(BaseModel):
    personal_target: Optional[float] = None
    notes: Optional[str] = None
    status: Optional[ParticipationStatus] = None

class UserChallengeRead(BaseModel):
    id: str
    user_id: str
    username: str
    challenge_id: str
    status: ParticipationStatus
    joined_at: datetime
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    updated_at: datetime
    progress: ChallengeProgress
    personal_target: Optional[float] = None
    notes: Optional[str] = None
    earned_points: int
    earned_badges: List[str]
    best_streak: int
    current_streak: int
    
    # Challenge info
    challenge_title: str
    challenge_type: ChallengeType
    challenge_difficulty: ChallengeDifficulty

class UserChallengeListResponse(BaseModel):
    user_challenges: List[UserChallengeRead]
    total_count: int
    skip: int
    limit: int